lib/Wikia/src/Service/User/Attributes/UserAttributes.php

```diff
@@ -64,38 +64,77 @@ private function loadAttributes( $userId ) {
 	 * @param string $userId
 	 * @param Attribute $attribute
 	 */
-	public function setAttribute( $userId, $attribute ) {
+	public function setAttribute( $userId, Attribute $attribute ) {
 		if ( $this->isAnonUser( $userId ) ) {
 			return;
 		}
 
-		$this->loadAttributes( $userId );
-		$this->setAttributeInService( $userId, $attribute );
-		$this->setAttributeInCache( $userId, $attribute );
+		if ( $this->attributeValueAlreadySet( $userId, $attribute ) ) {
+			return;
+		}
 
+		$this->setAttributeInCache( $userId, $attribute );
+		$this->setAttributeInService( $userId, $attribute );
 	}
 
 	private function isAnonUser( $userId ) {
 		return $userId === 0;
 	}
 
+	private function attributeValueAlreadySet( $userId, Attribute $attribute ) {
+		return $this->getAttribute( $userId, $attribute->getName() ) == $attribute->getValue();
+	}
+
 	/**
 	 * @param $userId
-	 * @param Attribute $attribute
+	 * @param array $attributes
 	 */
-	private function setAttributeInService( $userId, $attribute ) {
-		$this->attributeService->set( $userId, $attribute );
+	public function setAttributesInCache( $userId, array $attributes ) {
+		foreach ( $attributes as $attributeKey => $attributeValue ) {
+			$this->setAttributeInCache( $userId, new Attribute( $attributeKey, $attributeValue ) );
+		}
 	}
 
 	/**
 	 * @param $userId
 	 * @param Attribute $attribute
 	 */
-	private function setAttributeInCache( $userId, $attribute ) {
+	private function setAttributeInCache( $userId, Attribute $attribute ) {
 		$this->attributes[$userId][$attribute->getName()] = $attribute->getValue();
 	}
 
-	public function deleteAttribute( $userId, $attribute ) {
+	/**
+	 * @param $userId
+	 * @param Attribute $attribute
+	 */
+	private function setAttributeInService( $userId, $attribute ) {
+		$this->attributeService->set( $userId, $attribute );
+	}
+
+	public function deleteAttribute( $userId, Attribute $attribute ) {
+		if ( $this->isAnonUser( $userId ) ) {
```

Comment:
Personally I like the explicitness of keeping the 2 separate. Having `attributeNotSetForUser` check if `isAnonUser` works, but they do feel distinct to me. Either you're anon, or you're a logged in user and we're checking if you have the attribute set. It feels more clear this way.